{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "mount_file_id": "1l9SqHXMUsEai0hVZywWJyArcMTUnbyF8",
      "authorship_tag": "ABX9TyPruIvP3LDUZRkLFqoqnF0a",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/tiyash-dutta24-git/Tiyash-LS-ML-assignments/blob/main/You_Only_Look_Once.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# prompt: unzip the zipped train, test and valid datasets\n",
        "\n",
        "!unzip /content/train-20240711T161612Z-001.zip\n",
        "!unzip /content/test-20240711T161609Z-001.zip\n",
        "!unzip /content/valid-20240711T161617Z-001.zip\n"
      ],
      "metadata": {
        "id": "LcOGtmpRmGdF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# prompt: import a pre-trained yolov8 model, use it to train on train data, test on test data and validate on valid data. data.yml file is given - /content/data.yaml. After that generate confusion matrix and result.png, everything correctly.\n",
        "\n",
        "import ultralytics\n",
        "model = ultralytics.YOLO('/content/yolov8n-cls.pt')  # load a pretrained YOLOv8x model\n",
        "\n",
        "# Train the model on the train dataset\n",
        "!yolo task=classify mode=train data=/content/data.yaml model=yolov8n-cls.pt epochs=1 imgsz=640\n",
        "\n",
        "# Test the model on the test dataset\n",
        "!yolo task=classify mode=test data=/content/data.yaml model=yolov8n-cls.pt\n",
        "\n",
        "# Validate the model on the valid dataset\n",
        "!yolo task=classify mode=val data=/content/data.yaml model=yolov8n-cls.pt\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lBLFlngDIUz_",
        "outputId": "8bed6a74-5a15-47e2-d078-ee81fce9dd18"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Ultralytics YOLOv8.2.54 üöÄ Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)\n",
            "\u001b[34m\u001b[1mengine/trainer: \u001b[0mtask=classify, mode=train, model=yolov8n-cls.pt, data=/content/data.yaml, epochs=1, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train10, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/classify/train10\n",
            "\n",
            "Dataset not found ‚ö†Ô∏è, missing path /content/data.yaml, attempting download...\n",
            "Dataset download success ‚úÖ (0.0s), saved to \u001b[1m/content/data.yaml\u001b[0m\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py\", line 520, in get_dataset\n",
            "    data = check_cls_dataset(self.args.data)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/data/utils.py\", line 401, in check_cls_dataset\n",
            "    names = [x.name for x in (data_dir / \"train\").iterdir() if x.is_dir()]  # class names list\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/data/utils.py\", line 401, in <listcomp>\n",
            "    names = [x.name for x in (data_dir / \"train\").iterdir() if x.is_dir()]  # class names list\n",
            "  File \"/usr/lib/python3.10/pathlib.py\", line 1017, in iterdir\n",
            "    for name in self._accessor.listdir(self):\n",
            "NotADirectoryError: [Errno 20] Not a directory: '/content/data.yaml/train'\n",
            "\n",
            "The above exception was the direct cause of the following exception:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/yolo\", line 8, in <module>\n",
            "    sys.exit(entrypoint())\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py\", line 708, in entrypoint\n",
            "    getattr(model, mode)(**overrides)  # default args from model\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py\", line 644, in train\n",
            "    self.trainer = (trainer or self._smart_load(\"trainer\"))(overrides=args, _callbacks=self.callbacks)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/models/yolo/classify/train.py\", line 38, in __init__\n",
            "    super().__init__(cfg, overrides, _callbacks)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py\", line 132, in __init__\n",
            "    self.trainset, self.testset = self.get_dataset()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py\", line 531, in get_dataset\n",
            "    raise RuntimeError(emojis(f\"Dataset '{clean_url(self.args.data)}' error ‚ùå {e}\")) from e\n",
            "RuntimeError: Dataset '/content/data.yaml' error ‚ùå [Errno 20] Not a directory: '/content/data.yaml/train'\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/yolo\", line 8, in <module>\n",
            "    sys.exit(entrypoint())\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py\", line 649, in entrypoint\n",
            "    raise ValueError(f\"Invalid 'mode={mode}'. Valid modes are {MODES}.\\n{CLI_HELP_MSG}\")\n",
            "ValueError: Invalid 'mode=test'. Valid modes are {'train', 'track', 'val', 'benchmark', 'export', 'predict'}.\n",
            "\n",
            "    Arguments received: ['yolo', 'task=classify', 'mode=test', 'data=/content/data.yaml', 'model=yolov8n-cls.pt']. Ultralytics 'yolo' commands use the following syntax:\n",
            "\n",
            "        yolo TASK MODE ARGS\n",
            "\n",
            "        Where   TASK (optional) is one of {'obb', 'classify', 'segment', 'pose', 'detect'}\n",
            "                MODE (required) is one of {'train', 'track', 'val', 'benchmark', 'export', 'predict'}\n",
            "                ARGS (optional) are any number of custom 'arg=value' pairs like 'imgsz=320' that override defaults.\n",
            "                    See all ARGS at https://docs.ultralytics.com/usage/cfg or with 'yolo cfg'\n",
            "\n",
            "    1. Train a detection model for 10 epochs with an initial learning_rate of 0.01\n",
            "        yolo train data=coco8.yaml model=yolov8n.pt epochs=10 lr0=0.01\n",
            "\n",
            "    2. Predict a YouTube video using a pretrained segmentation model at image size 320:\n",
            "        yolo predict model=yolov8n-seg.pt source='https://youtu.be/LNwODJXcvt4' imgsz=320\n",
            "\n",
            "    3. Val a pretrained detection model at batch-size 1 and image size 640:\n",
            "        yolo val model=yolov8n.pt data=coco8.yaml batch=1 imgsz=640\n",
            "\n",
            "    4. Export a YOLOv8n classification model to ONNX format at image size 224 by 128 (no TASK required)\n",
            "        yolo export model=yolov8n-cls.pt format=onnx imgsz=224,128\n",
            "\n",
            "    5. Explore your datasets using semantic search and SQL with a simple GUI powered by Ultralytics Explorer API\n",
            "        yolo explorer\n",
            "    \n",
            "    6. Streamlit real-time object detection on your webcam with Ultralytics YOLOv8\n",
            "        yolo streamlit-predict\n",
            "        \n",
            "    7. Run special commands:\n",
            "        yolo help\n",
            "        yolo checks\n",
            "        yolo version\n",
            "        yolo settings\n",
            "        yolo copy-cfg\n",
            "        yolo cfg\n",
            "\n",
            "    Docs: https://docs.ultralytics.com\n",
            "    Community: https://community.ultralytics.com\n",
            "    GitHub: https://github.com/ultralytics/ultralytics\n",
            "    \n",
            "Ultralytics YOLOv8.2.54 üöÄ Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)\n",
            "YOLOv8n-cls summary (fused): 73 layers, 2715880 parameters, 0 gradients, 4.3 GFLOPs\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/yolo\", line 8, in <module>\n",
            "    sys.exit(entrypoint())\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py\", line 708, in entrypoint\n",
            "    getattr(model, mode)(**overrides)  # default args from model\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py\", line 517, in val\n",
            "    validator(model=self.model)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py\", line 115, in decorate_context\n",
            "    return func(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/engine/validator.py\", line 143, in __call__\n",
            "    self.data = check_det_dataset(self.args.data)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/data/utils.py\", line 279, in check_det_dataset\n",
            "    data = yaml_load(file, append_filename=True)  # dictionary\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ultralytics/utils/__init__.py\", line 380, in yaml_load\n",
            "    data = yaml.safe_load(s) or {}  # always return a dict (yaml.safe_load() may return None for empty files)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/__init__.py\", line 125, in safe_load\n",
            "    return load(stream, SafeLoader)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/__init__.py\", line 81, in load\n",
            "    return loader.get_single_data()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/constructor.py\", line 49, in get_single_data\n",
            "    node = self.get_single_node()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/composer.py\", line 36, in get_single_node\n",
            "    document = self.compose_document()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/composer.py\", line 58, in compose_document\n",
            "    self.get_event()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/parser.py\", line 118, in get_event\n",
            "    self.current_event = self.state()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/parser.py\", line 193, in parse_document_end\n",
            "    token = self.peek_token()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/scanner.py\", line 129, in peek_token\n",
            "    self.fetch_more_tokens()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/scanner.py\", line 223, in fetch_more_tokens\n",
            "    return self.fetch_value()\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/yaml/scanner.py\", line 577, in fetch_value\n",
            "    raise ScannerError(None, None,\n",
            "yaml.scanner.ScannerError: mapping values are not allowed here\n",
            "  in \"<unicode string>\", line 7, column 9:\n",
            "    roboflow:\n",
            "            ^\n"
          ]
        }
      ]
    }
  ]
}